src/main.cpp

```diff
@@ -6,142 +6,96 @@
+
+
+	// read default action
 
-	// look for default rule in lua parser script
-	bool foundDefault = true;
-	lua_getglobal(L.get(), "default");
-	if(!lua_isfunction(L.get(), -1)){
-		foundDefault = false;
+	lua_pushvalue(L, 3); // copy the function handle to top
+	string defaultFuncName = ptr2str(lua_topointer(L, -1)); // use the address of the function as its name
+	lua_setglobal(L, defaultFuncName.c_str()); // store function in lua globals (pops from stack)
+
+	auto rules = pegParser->get_rule_names();
+	for(auto& rule:rules){ // register default action for all PEG definitions
+		registerReductionRule(pegParser, L, rule, defaultFuncName);
 	}
 
-	// assign in-lua-defined rules
-	for(const auto& rule:rules){
-		string funcName = rule.c_str();
-		lua_getglobal(L.get(), funcName.c_str()); // put pointer to function on stack
-		if(!lua_isfunction(L.get(), -1)){
-			if(foundDefault){funcName = "default";}
-			else{funcName = "";}
-		}
 
-		if(funcName != ""){ // use the specified function or the default function from the lua script
-			parser[rule.c_str()] = [rule, funcName](const SemanticValues& sv, any&){
-
-				// push input parameters on stack
-				const lua::value params = lua::newtable();
-				params["choice"] = sv.choice();
-				params["column"] = sv.line_info().second;
-				params["line"] = sv.line_info().first;
-				params["matched"] = StringPtr(sv.c_str(), sv.length());
-				params["rule"] = rule;
-
-				const lua::value subnodes = lua::newtable();
-				for (size_t i = 0; i != sv.size(); ++i){
-					subnodes[1 + i] = sv[i].get<lua::value>();
-				}
-				params["subnodes"] = subnodes;
-
-				const lua::value tokens = lua::newtable();
-				for (size_t i = 0; i != sv.tokens.size(); ++i) {
-					tokens[1 + i] = StringPtr(sv.tokens[i].first, sv.tokens[i].second);
-				}
-				params["tokens"] = tokens;
-
-				// find function
-				const lua::value func(lua::globals()[funcName.c_str()]);
-
-				// call lua function and return lua value.
-				return func(params);
-			};
-		}
-		else{ // function not found, no default function in lua script -> just return last matched token
-			parser[rule.c_str()] = [&L](const SemanticValues& sv, any&){
-				if(sv.tokens.size() == 0){
-					lua_pushlstring(L.get(), sv.c_str(), sv.length()); // no tokens, return matched string
-				}
-				else {
-					lua_pushlstring(L.get(), sv.tokens.back().first, sv.tokens.back().second); // return last token
-				}
-
-				// return lua value
-				return lua::value();
-			};
-		}
+	// read action table ( https://stackoverflow.com/questions/6137684/iterate-through-lua-table )
+	lua_pushvalue(L, 2);
+	lua_pushnil(L);
+	while (lua_next(L, -2))
+	{
+		lua_pushvalue(L, -2);
+		string rule = lua_tostring(L, -1); // = key
+		string funcName = ptr2str(lua_topointer(L, -2)); // = value (use the address of the function as its name)
+		lua_pop(L, 1); // pop key
+		lua_setglobal(L, funcName.c_str()); // store function in lua globals (pops from stack)
```

Comment:
Note that this does *not* store the function in Lua's registry (as suggested by a comment in `registerReductionRule`) but rather as a global variable which is also visible to Lua code. Storing the value in the registry instead has the advantage that Lua code doesn't see arbitrary global variables appear apparently from nowhere and reduces the risk of inadvertently changing them.

Reply:
In that comment I was referring to the... glossary where lua's globals are registered. I was not aware there is yet another, more registry-esque registry which is more suitable. Thanks for pointing that out.

Reply:
`lua_getglobal()` and `lua_setglobal()` merely modify the table of global variables. This table can also be accessed using `lua_pushglobaltable()`.

In contrast, the pseudo index [`LUA_REGISTRYINDEX`](https://www.lua.org/manual/5.3/manual.html#4.5) refers to a table that is only accessible to the host.